include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+NBL_CONSTEXPR uint32_t BucketsPerThread = ceil((float) BucketCount / WorkgroupSize);
+
+groupshared uint32_t prefixScratch[BucketCount];
+
+struct ScratchProxy
+{
+    uint32_t get(const uint32_t ix)
+    {
+        return prefixScratch[ix];
+    }
+    void set(const uint32_t ix, const uint32_t value)
+    {
+        prefixScratch[ix] = value;
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
+
+template<typename KeyAccessor, typename ValueAccessor, typename ScratchAccessor>
+struct counting
+{
+    void histogram(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            int j = index + i * WorkgroupSize + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            nbl::hlsl::glsl::atomicAdd(sdata[k - data.minimum], (uint32_t) 1);
+        }
+
+        nbl::hlsl::glsl::barrier();
+
+        uint32_t sum = 0;
+        uint32_t scan_sum = 0;
+
+        for (int i = 0; i < BucketsPerThread; i++)
+        {
+            sum = nbl::hlsl::workgroup::exclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
+            template __call <ScratchProxy>
+            (sdata[WorkgroupSize * i + tid], arithmeticAccessor);
```

Comment:
ok here's the trick, a workgroup prefix sum or reduction uses the bottom/beginning of your scratch accessor and at most 1/4 uint per invocation in a workgroup

Meaning that you only need to preload the first Workgroup-sized chunk of the histogram scratch into a local variable before you enter the loop, issue a barrier (cause you alias the scratches), then proceed to do an exclusive scan from the local variables


```
// wait for the histogramming to finish
scratchAccessor.workgroupExecutionAndMemoryBarrier();

const K preload = scratchAccessor.get(tid);
// because first chunk of histogram and workgroup scan scratch are aliased
scratchAccessor.workgroupExecutionAndMemoryBarrier();
uint32_t sum = inclusive_scan(preload,scratchAccessor);
atomicAddIntoGlobal(tid,sum);
for (int i=1; ...) // start loop at iteration 1
{
  // no if statement about the last iteration needed
  if (tid==WorkgroupSize-1)
  {
     const uint32_t nextBlockStart = WorkgroupSize*i;
     scratchAccessor.set(nextBlockStart,scratchAccessor.get(nextBlockStart)+sum);
  }
  // propagate last block tail to next block head and protect against subsequent scans stepping on each other's toes
  scratchAccessor.workgroupExecutionAndMemoryBarrier();
  // no aliasing anymore
  const uint32_t index = WorkgroupSize*i+tid;
  sum = inclusive_scan(scratchAccessor.get(index),scratchAccessor);
  atomicAddIntoGlobal(index,sum);
}
```

P.s. I believe we talked that going back to inclusive sum is easier on the workload.

Reply:
Done

Reply:
btw I gave comments in my pseudocode, would be useful to have them in the final code